### ISSUE TITLE
need help with high app crash rate

### ISSUE THREAD

**Issue by littlehome-eugene** [transistorsoft/react-native-background-geolocation#969]
## Your Environment

* Plugin version: "react-native-background-geolocation": "^3.2.0",
* Platform: iOS or Android android
* OS version: mac high sierra
* Device manufacturer / model: galaxy s10e
* React Native version (`react-native -v`):
react-native-cli: 2.0.1
react-native: 0.59.10
* Plugin config
```javascript <!-- Syntax-highlighting:  paste your code below -->
    let config = {
      reset: false,
      stopTimeout: 1,
      debug: DEBUG && __DEV__,
      logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
      foregroundService: true,
      autoSync: false,
      url: "",
      //            url: TRACKER_HOST + this.state.username,
      stopOnTerminate: false,
      startOnBoot: true,
      heartbeatInterval: 60,
      enableHeadless: true,
      params: BackgroundGeolocation.transistorTrackerParams(DeviceInfo),
      maxDaysToPersist: 14,
      geofenceProximityRadius: 3000,
      disableLocationAuthorizationAlert: true
      /*
       * disableLocationAuthorizationAlert
       Disables automatic authorization alert when plugin detects the user has disabled location authorization.
       You will be responsible for handling disabled location authorization
       by listening to the providerchange event.
      */
    }

    BackgroundGeolocation.ready(config, (state:State) => {
      if (state.schedule && state.schedule.length > 0) {
        BackgroundGeolocation.startSchedule();
      }
      this.setState({
        enabled: state.enabled,
        isMoving: state.isMoving,
        followsUserLocation: state.enabled,
        showsUserLocation: state.enabled,
        bgGeo: state
      });
    }, (error:string) => {
    });

# more code

if(Platform.OS === 'android'){
   // https:github.com/transistorsoft/react-native-background-geolocation/wiki/Android-Headless-Mode
   let BackgroundFetch = require("react-native-background-fetch").default
   let BackgroundGeolocation = require('react-native-background-geolocation').default
   let bgMessaging = require('site_common/react/helper/fix-no-task-registered-key-background-fetch').default
   AppRegistry.registerHeadlessTask('RNFirebaseBackgroundMessage', () => bgMessaging);//https:github.com/transistorsoft/react-native-background-fetch/issues/101
   // https:github.com/transistorsoft/react-native-background-fetch/issues/181
   let BackgroundFetchHeadlessTask = async (event) => {
     BackgroundFetch.stop();
     BackgroundFetch.finish();
   }
   BackgroundFetch.registerHeadlessTask(BackgroundFetchHeadlessTask);
   let BackgroundGeolocationHeadlessTask = async (event) => {
   }
   BackgroundGeolocation.registerHeadlessTask(BackgroundGeolocationHeadlessTask);
 }

 componentDidMount() {
    this.failRequestPermission()
   }
   _handleAppStateChange(state:any) {
     Permissions.check('location', { type: 'always' }).then(response => {
       if(response != 'authorized') {
         this.failRequestPermission()
       }
     })
   }
   failRequestPermission() {
     BackgroundGeolocation.stop();
     // Clear markers, polyline, geofences, stationary-region
     this.setState({
       enabled: false,
       isMoving: false,
       showsUserLocation: false,
       followsUserLocation: false,
       stationaryRadius: 0,
       stationaryLocation: {
         timestamp: '',
         latitude: 0,
         longitude: 0
       }
     });
     BackgroundGeolocation.removeListeners();
   }

```

## Actual Behavior

too many crashes in play console

Crash due to bg-geolocation is about 2% of all users.  
It's hard to fix because we can't reproduce it ourselves..

(if we knew how to reproduce this, it'd be much easier :()

## Steps to Reproduce

1.
2.
3.
4.

## Context
Track you location
when user enter a geofence, report it to our server

## Debug logs

<details><summary>Logs</summary>

``` <!-- Syntax highlighting:  DO NOT REMOVE -->
# Crashlytics - plaintext stacktrace downloaded by zibann dev at Tue, 04 Feb 2020 08:38:41 GMT
# URL: https://fabric.io/zibann-devs-projects/android/apps/com.zibann.littlehome/issues/5be5841ff8b88c296332ef46/sessions/latest
# Organization: zibann dev’s Projects
# Platform: android
# Application: Little Home
# Version: 2.11.63 (273)
# Bundle Identifier: com.zibann.littlehome
# Issue ID: 5be5841ff8b88c296332ef46
# Session ID: 5E392821010800013625AE5DD38A37A2_DNE_0_v2
# Date: 2020-02-04T08:15:00Z
# OS Version: 9
# Device: SM-N960N
# RAM Free: 45.6%
# Disk Free: 83.8%
#0. Crashed: mqt_native_modules
       at com.facebook.react.modules.core.ExceptionsManagerModule.showOrThrowError + 54(ExceptionsManagerModule.java:54)
       at com.facebook.react.modules.core.ExceptionsManagerModule.reportFatalException + 38(ExceptionsManagerModule.java:38)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.facebook.react.bridge.JavaMethodWrapper.invoke + 372(JavaMethodWrapper.java:372)
       at com.facebook.react.bridge.JavaModuleWrapper.invoke + 158(JavaModuleWrapper.java:158)
       at com.facebook.react.bridge.queue.NativeRunnable.run(NativeRunnable.java)
       at android.os.Handler.handleCallback + 873(Handler.java:873)
       at android.os.Handler.dispatchMessage + 99(Handler.java:99)
       at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage + 29(MessageQueueThreadHandler.java:29)
       at android.os.Looper.loop + 214(Looper.java:214)
       at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run + 232(MessageQueueThreadImpl.java:232)
       at java.lang.Thread.run + 764(Thread.java:764)
--
Fatal Exception: com.facebook.react.common.JavascriptException: No task registered for key BackgroundGeolocation, stack:
startHeadlessTask@337:2882
value@36:3311
<unknown>@36:822
value@36:2565
value@36:794
value@-1
       at com.facebook.react.modules.core.ExceptionsManagerModule.showOrThrowError + 54(ExceptionsManagerModule.java:54)
       at com.facebook.react.modules.core.ExceptionsManagerModule.reportFatalException + 38(ExceptionsManagerModule.java:38)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.facebook.react.bridge.JavaMethodWrapper.invoke + 372(JavaMethodWrapper.java:372)
       at com.facebook.react.bridge.JavaModuleWrapper.invoke + 158(JavaModuleWrapper.java:158)
       at com.facebook.react.bridge.queue.NativeRunnable.run(NativeRunnable.java)
       at android.os.Handler.handleCallback + 873(Handler.java:873)
       at android.os.Handler.dispatchMessage + 99(Handler.java:99)
       at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage + 29(MessageQueueThreadHandler.java:29)
       at android.os.Looper.loop + 214(Looper.java:214)
       at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run + 232(MessageQueueThreadImpl.java:232)
       at java.lang.Thread.run + 764(Thread.java:764)
#0. Crashed: mqt_native_modules
       at com.facebook.react.modules.core.ExceptionsManagerModule.showOrThrowError + 54(ExceptionsManagerModule.java:54)
       at com.facebook.react.modules.core.ExceptionsManagerModule.reportFatalException + 38(ExceptionsManagerModule.java:38)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.facebook.react.bridge.JavaMethodWrapper.invoke + 372(JavaMethodWrapper.java:372)
       at com.facebook.react.bridge.JavaModuleWrapper.invoke + 158(JavaModuleWrapper.java:158)
       at com.facebook.react.bridge.queue.NativeRunnable.run(NativeRunnable.java)
       at android.os.Handler.handleCallback + 873(Handler.java:873)
       at android.os.Handler.dispatchMessage + 99(Handler.java:99)
       at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage + 29(MessageQueueThreadHandler.java:29)
       at android.os.Looper.loop + 214(Looper.java:214)
       at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run + 232(MessageQueueThreadImpl.java:232)
       at java.lang.Thread.run + 764(Thread.java:764)
#1. GoogleApiHandler
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next + 326(MessageQueue.java:326)
       at android.os.Looper.loop + 181(Looper.java:181)
       at android.os.HandlerThread.run + 65(HandlerThread.java:65)
#2. Queue
       at java.lang.Object.wait(Object.java)
       at java.lang.Thread.parkFor$ + 2137(Thread.java:2137)
       at sun.misc.Unsafe.park + 358(Unsafe.java:358)
       at java.util.concurrent.locks.LockSupport.park + 190(LockSupport.java:190)
       at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await + 2059(AbstractQueuedSynchronizer.java:2059)
       at java.util.concurrent.PriorityBlockingQueue.take + 548(PriorityBlockingQueue.java:548)
       at io.fabric.sdk.android.services.concurrency.DependencyPriorityBlockingQueue.performOperation + 197(DependencyPriorityBlockingQueue.java:197)
       at io.fabric.sdk.android.services.concurrency.DependencyPriorityBlockingQueue.get + 236(DependencyPriorityBlockingQueue.java:236)
       at io.fabric.sdk.android.services.concurrency.DependencyPriorityBlockingQueue.take + 65(DependencyPriorityBlockingQueue.java:65)
       at io.fabric.sdk.android.services.concurrency.DependencyPriorityBlockingQueue.take + 46(DependencyPriorityBlockingQueue.java:46)
       at java.util.concurrent.ThreadPoolExecutor.getTask + 1092(ThreadPoolExecutor.java:1092)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1152(ThreadPoolExecutor.java:1152)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 641(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run + 764(Thread.java:764)
#3. Segment-SegmentDispatcher
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next + 326(MessageQueue.java:326)
       at android.os.Looper.loop + 181(Looper.java:181)
       at android.os.HandlerThread.run + 65(HandlerThread.java:65)
#4. HybridData DestructorThread
       at java.lang.Object.wait(Object.java)
       at java.lang.Object.wait + 422(Object.java:422)
       at java.lang.ref.ReferenceQueue.remove + 188(ReferenceQueue.java:188)
       at java.lang.ref.ReferenceQueue.remove + 209(ReferenceQueue.java:209)
       at com.facebook.jni.DestructorThread$1.run + 63(DestructorThread.java:63)
#5. mqt_native_modules
       at java.lang.Object.wait(Object.java)
       at java.lang.Thread.parkFor$ + 2137(Thread.java:2137)
       at sun.misc.Unsafe.park + 358(Unsafe.java:358)
       at java.util.concurrent.locks.LockSupport.park + 190(LockSupport.java:190)
       at java.util.concurrent.FutureTask.awaitDone + 450(FutureTask.java:450)
       at java.util.concurrent.FutureTask.get + 192(FutureTask.java:192)
       at com.crashlytics.android.core.CrashlyticsExecutorServiceWrapper.executeSyncLoggingException + 47(CrashlyticsExecutorServiceWrapper.java:47)
       at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler.uncaughtException + 235(CrashlyticsUncaughtExceptionHandler.java:235)
       at com.transistorsoft.locationmanager.adapter.BackgroundGeolocation$e.uncaughtException + 106(:106)
       at io.sentry.SentryUncaughtExceptionHandler.uncaughtException + 64(SentryUncaughtExceptionHandler.java:64)
       at java.lang.ThreadGroup.uncaughtException + 1068(ThreadGroup.java:1068)
       at java.lang.ThreadGroup.uncaughtException + 1063(ThreadGroup.java:1063)
       at java.lang.Thread.dispatchUncaughtException + 1955(Thread.java:1955)
#6. Queue
       at java.lang.Object.wait(Object.java)
       at java.lang.Thread.parkFor$ + 2137(Thread.java:2137)
       at sun.misc.Unsafe.park + 358(Unsafe.java:358)
       at java.util.concurrent.locks.LockSupport.park + 190(LockSupport.java:190)
       at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await + 2059(AbstractQueuedSynchronizer.java:2059)
       at java.util.concurrent.PriorityBlockingQueue.take + 548(PriorityBlockingQueue.java:548)
       at io.fabric.sdk.android.services.concurrency.DependencyPriorityBlockingQueue.performOperation + 197(DependencyPriorityBlockingQueue.java:197)
       at io.fabric.sdk.android.services.concurrency.DependencyPriorityBlockingQueue.get + 236(DependencyPriorityBlockingQueue.java:236)
       at io.fabric.sdk.android.services.concurrency.DependencyPriorityBlockingQueue.take + 65(DependencyPriorityBlockingQueue.java:65)
       at io.fabric.sdk.android.services.concurrency.DependencyPriorityBlockingQueue.take + 46(DependencyPriorityBlockingQueue.java:46)
       at java.util.concurrent.ThreadPoolExecutor.getTask + 1092(ThreadPoolExecutor.java:1092)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1152(ThreadPoolExecutor.java:1152)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 641(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run + 764(Thread.java:764)
#7. mqt_js
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next + 326(MessageQueue.java:326)
       at android.os.Looper.loop + 181(Looper.java:181)
       at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run + 232(MessageQueueThreadImpl.java:232)
       at java.lang.Thread.run + 764(Thread.java:764)
#8. GoogleApiHandler
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next + 326(MessageQueue.java:326)
       at android.os.Looper.loop + 181(Looper.java:181)
       at android.os.HandlerThread.run + 65(HandlerThread.java:65)
#9. Segment-Stats
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next + 326(MessageQueue.java:326)
       at android.os.Looper.loop + 181(Looper.java:181)
       at android.os.HandlerThread.run + 65(HandlerThread.java:65)
#10. AsyncTask #3
       at java.lang.Object.wait(Object.java)
       at java.lang.Thread.parkFor$ + 2137(Thread.java:2137)
       at sun.misc.Unsafe.park + 358(Unsafe.java:358)
       at java.util.concurrent.locks.LockSupport.parkNanos + 230(LockSupport.java:230)
       at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos + 2101(AbstractQueuedSynchronizer.java:2101)
       at java.util.concurrent.LinkedBlockingQueue.poll + 467(LinkedBlockingQueue.java:467)
       at java.util.concurrent.ThreadPoolExecutor.getTask + 1091(ThreadPoolExecutor.java:1091)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1152(ThreadPoolExecutor.java:1152)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 641(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run + 764(Thread.java:764)
#11. Okio Watchdog
       at java.lang.Object.wait(Object.java)
       at com.android.okhttp.okio.AsyncTimeout.awaitTimeout + 311(AsyncTimeout.java:311)
       at com.android.okhttp.okio.AsyncTimeout.access$000 + 40(AsyncTimeout.java:40)
       at com.android.okhttp.okio.AsyncTimeout$Watchdog.run + 286(AsyncTimeout.java:286)
#12. Crashlytics Exception Handler1
       at dalvik.system.VMStack.getThreadStackTrace(VMStack.java)
       at java.lang.Thread.getStackTrace + 1538(Thread.java:1538)
       at java.lang.Thread.getAllStackTraces + 1588(Thread.java:1588)
       at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler.writeSessionEvent + 1024(CrashlyticsUncaughtExceptionHandler.java:1024)
       at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler.writeFatal + 766(CrashlyticsUncaughtExceptionHandler.java:766)
       at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler.handleUncaughtException + 262(CrashlyticsUncaughtExceptionHandler.java:262)
       at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler.access$100 + 55(CrashlyticsUncaughtExceptionHandler.java:55)
       at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler$5.call + 238(CrashlyticsUncaughtExceptionHandler.java:238)
       at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler$5.call + 235(CrashlyticsUncaughtExceptionHandler.java:235)
       at java.util.concurrent.FutureTask.run + 266(FutureTask.java:266)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1167(ThreadPoolExecutor.java:1167)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 641(ThreadPoolExecutor.java:641)
       at io.fabric.sdk.android.services.common.ExecutorUtils$1$1.onRun + 75(ExecutorUtils.java:75)
       at io.fabric.sdk.android.services.common.BackgroundPriorityRunnable.run + 30(BackgroundPriorityRunnable.java:30)
       at java.lang.Thread.run + 764(Thread.java:764)
#13. ConnectivityThread
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next + 326(MessageQueue.java:326)
       at a...
```

</details>

**Comment by christocracy** [comment]
Because you probably blindly copy/pasted code `enableHeadless: true` and didn't configure your app for [Android Headless Mode](https://github.com/transistorsoft/react-native-background-geolocation/wiki/Android-Headless-Mode).

**Comment by littlehome-eugene** [comment]
From your prev answer: 

> You can't "disable" the plugin by merely commenting out the Javascript code. You need to tell the plugin to BackgroundGeolocation.stop() before you comment-out the code. The plugin operates independently of your Javascript and keeps track of its state between reboots. Once the plugin is enabled (via BackgroundGeolocation.start()), it will be enabled forever until you BackgroundGeolocation.stop().


Does this mean, it could be the very reason why we are getting the crash? 

Because we commented out the headless code without turning it off, that's why we are getting the crash?

**Comment by littlehome-eugene** [comment]
> Because you probably blindly copy/pasted code enableHeadless: true and didn't configure your app for Android Headless Mode.

We pasted our code, we didn't blindly copy-paste , although I don't see there's any that we can do any differently from the doc you mentioned

**Comment by littlehome-eugene** [comment]
We have the following code, and still receiving crash report.  

We want to turn off the plugin ( because we can't fix the crash)
We want to disable the notification 

1. headeless task is registered
2. calling `BackgroundGelocation.stop()` on app start
3. we are not calling `ready()` function (commented out)

```
***index.js
    let BackgroundGeolocationHeadlessTask = async (event) => {
    }
    BackgroundGeolocation.registerHeadlessTask(BackgroundGeolocationHeadlessTask)
***background-geolocaiton.tsx
  componentDidMount() {
  BackgroundGeolocation.stop();
  BackgroundGeolocaiton.removeListener();
}
```

**Comment by christocracy** [comment]
```
BackgroundGeolocaiton.removeListener();
```
That should be #removeListeners (plural)

**Comment by littlehome-eugene** [comment]
sorry we had `removeListeners` it was copy/paste mistake.

So is there anything else we can do to stop the crash?

**Comment by christocracy** [comment]
I have no idea what you’ve done.  Remove the plugin and start over.

**Comment by littlehome-eugene** [comment]
I wish I could just start over

we couldn't remove it, because then there are people who get the notifications (Location service activate) We need to to call `BackgroundGeolocation.stop();` for them.

Does `BackgroundGelocation.stop()` stops the service (?) that calls the registered headless task?

I think the following code needs to be undone to stop our crash. how to do that?
Would we have a better shot calling `ready()` with the approapriate options to turn it off?

 ```
   BackgroundGeolocation.ready({
      enableHeadless: true,
      stopOnTerminate: false,  // <-- required for Headless JS
      .
      .
      .
    }, (state) => {
      console.log('- Configure success');
    });
```

**Comment by christocracy** [comment]
If I were you, I’d enableHeadless: false and do more testing before releasing.

When callback to #ready executes, call #stop.

**Comment by littlehome-eugene** [comment]
Thanks for suggestion
Suppose

We comment out all the code
if a user gets the `location activated` notification, let him uninstall /install the app.
would he still get `location activated` notification?

**Comment by christocracy** [comment]
When the plugin’s #start / #stop methods act exactly like the power button on any electronic device.

When you #stop the plugin, it stays stopped — forever.  When you #start the plugin, it stays started forever (regardless if you comment out the JavaScript).  The plugin is essentially an autonomous, headless, pure native sub-application.

**Comment by littlehome-eugene** [comment]
what if a user uninstalls app. does it stop ? or does it continue runinning after uninstall / reinstall?

**Comment by christocracy** [comment]
The plugin is a merely a 3rd party library installed in your app.  When your app is uninstalled, of course the plugin no longer exists.

**Comment by littlehome-eugene** [comment]
Thanks for responding

I still don't understand why we are getting `Fatal Exception: com.facebook.react.common.JavascriptException: No task registered for key BackgroundGeolocation, stack:` when we have

`let BackgroundGeolocationHeadlessTask = async (event) => {
    }
    BackgroundGeolocation.registerHeadlessTask(BackgroundGeolocationHeadlessTask)`

We have another headless for background push notification which works fine, so I don't think we messed up registering the task.


So we decided to remove all the code and hope the crash go away
Some users might receive `location activated` notification, and we will inform them to reinstall. 

again, thanks for support, 
it must be our mistake somewhere if no other users experience the above error. but it's puzzling why we are getting it
